### PR TITLE
further improvements to registry API handling

### PR DESCRIPTION
**Pull request by tomerd**
motivation: improved registry support

changes:
* make use of availability information when persent
* improve handling of 404 responses
* add and improve tests

### CODE REVIEW

**Comment by tomerd** [comment]
needs #6170 first

**Comment by tomerd** [comment]
@swift-ci smoke test

File: Tests/PackageRegistryTests/RegistryClientTests.swift [hunk 1]
```diff
@@ -94,15 +94,46 @@ final class RegistryClientTests: XCTestCase {
...
-        let registryURL = "https://packages.example.com"
+        let registryURL = URL("https://packages.example.com")
         let identity = PackageIdentity.plain("mona.LinkedList")
         let releasesURL = URL("\(registryURL)/\(identity.registry!.scope)/\(identity.registry!.name)")
 
         let serverErrorHandler = ServerErrorHandler(
             method: .get,
             url: releasesURL,
-            errorCode: Int.random(in: 400 ..< 500),
+            errorCode: Int.random(in: 405 ..< 500),
```
Comment (yim-lee) [inline Tests/PackageRegistryTests/RegistryClientTests.swift:136]: 👍

**Review by yim-lee — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Pull request merged**